### PR TITLE
Refine resolution of overloaded constructor & field declarations

### PR DESCRIPTION
**Pull request by toinehartman**
1. Field with identical name/type signatures from distinct ADTs would incorrectly be classified as mutual overloads.
2. Constructor overloads where not detected correctly, and no constructor tests existed at all.

This PR refines the overload detection to solve both issues, and adds many tests.

### CODE REVIEW

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc [hunk 1]
```diff
@@ -237,12 +237,20 @@ set[loc] rascalGetOverloadedDefs(WorkspaceInfo ws, set[loc] defs, MayOverloadFun
       o scopeDefs                  // 3. Find definitions in the reached scope, and definitions within those definitions (transitively)
       ;
 
-    rel[loc from, loc to] defPaths = {};
+    rel[loc from, loc to] defPaths = fromDefPaths;
```
Comment (rodinaarssen) [inline rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc:240]: This seems to fix the `constructorId` branch as to not always leave `defPaths` empty. Did you fix something for which we did not have a test yet?

File: rascal-lsp/src/main/rascal/lang/rascal/tests/rename/Fields.rsc [hunk 2]
```diff
@@ -81,6 +81,19 @@ test bool commonKeywordFieldsSameType() = testRenameOccurrences({0, 1},
...
+", decls = "
+    'data D = d(int foo);
+    'data E = e(int foo);
+");
+
+test bool sameNameFieldsDisconnectedModules() = testRename({
+    byText("A", "data D = d(int foo);", {0})
+  , byText("B", "data E = e(int foo);")
+});
+
```
Comment (rodinaarssen) [inline rascal-lsp/src/main/rascal/lang/rascal/tests/rename/Fields.rsc:96]: Should we add a test for equal field names in different constructors of a single ADT?

**Review by rodinaarssen — COMMENTED** [review]
LGTM

File: rascal-lsp/src/main/rascal/lang/rascal/tests/rename/Fields.rsc [hunk 3]
```diff
@@ -81,6 +81,19 @@ test bool commonKeywordFieldsSameType() = testRenameOccurrences({0, 1},
...
+    'D x = d(8);
+    'int i = x.foo;
+", decls = "
+    'data D = d(int foo);
+    'data E = e(int foo);
+");
+
+test bool sameNameFieldsDisconnectedModules() = testRename({
+    byText("A", "data D = d(int foo);", {0})
+  , byText("B", "data E = e(int foo);")
```
Comment (rodinaarssen) [inline rascal-lsp/src/main/rascal/lang/rascal/tests/rename/Fields.rsc:94]:
```suggestion
  , byText("B", "data E = e(int foo);", {})
```

File: rascal-lsp/src/main/rascal/lang/rascal/tests/rename/Fields.rsc (hunk 2 above)
Comment (toinehartman) [reply rascal-lsp/src/main/rascal/lang/rascal/tests/rename/Fields.rsc:96]: We already have [this](https://github.com/usethesource/rascal-language-servers/pull/487/files#diff-93a989963a285dc6bd654287ba732805ffe9038a045761115ee7b35b86e3f00aL52-L56). Does this cover the cases that you are referring to?

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc (hunk 1 above)
Comment (toinehartman) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc:240]: I noticed that as well. I'll think about this and see if I can come up with a test that would have before this change.

Comment (toinehartman) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc]: Well, constructors were barely tested, and as you suspected, renaming them did not work as it should. Thanks for uncovering this. This led to an even more invasive rewrite of this overload detection code and adding many tests.

File: rascal-lsp/src/main/rascal/lang/rascal/tests/rename/Fields.rsc (hunk 3 above)
Comment (toinehartman) [reply rascal-lsp/src/main/rascal/lang/rascal/tests/rename/Fields.rsc]: This was an oversight that quickly broke the CI as well. Fixed now.

File: rascal-lsp/src/main/rascal/lang/rascal/tests/rename/Constructors.rsc [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+}
+module lang::rascal::tests::rename::Constructors
```
Comment (rodinaarssen) [inline rascal-lsp/src/main/rascal/lang/rascal/tests/rename/Constructors.rsc:27]: Nice amount of tests!

Constructor names can also occur in patterns (`foo(_, _) := c`, or as argument to a function) and after `is` (`if (c is foo);`). Consider adding tests for these situations

**Review by rodinaarssen — APPROVED** [review]
Left a remark, but looks good overall.

File: rascal-lsp/src/main/rascal/lang/rascal/tests/rename/Constructors.rsc (hunk 4 above)
Comment (toinehartman) [reply rascal-lsp/src/main/rascal/lang/rascal/tests/rename/Constructors.rsc:27]: Good one. Added in fe7845a.

**Pull request merged**